### PR TITLE
[Cocoa] font-optical-sizing:auto doesn't set the opsz variation axis for all fonts that need it

### PR DESCRIPTION
**Pull request by litherum**
#### b2e8181dbd7510819ee516e847f139fa7cb1b636
<pre>
[Cocoa] font-optical-sizing:auto doesn&apos;t set the opsz variation axis for all fonts that need it
<a href="https://bugs.webkit.org/show_bug.cgi?id=252552">https://bugs.webkit.org/show_bug.cgi?id=252552</a>
rdar://105662167

Reviewed by Alan Baradlay.

This is a follow-up from 260447@main. That previous patch went almost all the way toward fixing the
problem, but couldn&apos;t actually apply optical sizing to all text that requested it. The reason for
this is that the trigger we were using to enable optical sizing, `kCTFontOpticalSizeAttribute`,
enables the `opsz` variation axis, but also enables the `trak` table. Unfortunately, we have a lot
of tests which use fonts that don&apos;t support the optical sizing variation axis, but do have a `trak`
table. So, if we just naively started enabling `kCTFontOpticalSizeAttribute` for all text that
requests it, we&apos;ll get tons of test failures. It&apos;s not really great if we make tons of text on the
web look different, without it being a deliberate decision. So, the previous patch stopped just
short of enabling `kCTFontOpticalSizeAttribute` everywhere.

However, all fonts which *do* support the `opsz` variation axis should have it set automatically,
which is what 260447@main was aiming to address. We have the ability to *just* set the variation
axis without also setting `trak`, so that&apos;s what we need to do for fonts that request optical
sizing but for which we can&apos;t use `kCTFontOpticalSizeAttribute` because of compatibility.

So, we need 3 buckets:
- Disable optical sizing. This is triggered by font-optical-sizing:none. This maps to telling Core
      Text to use `kCTFontOpticalSizeAttribute`: `none`.
- Enable full optical sizing, including `trak`. This is triggered by
      text-rendering:optimizeLegibility. This maps to telling Core Text to use
      `kCTFontOpticalSizeAttribute`: `auto`.
- Enable optical sizing, but just the `opsz` variation, and not `trak`, for compatibility. This
      is what all situations not captured by the above use. This maps to telling Core Text to use
      `kCTFontVariationAttribute`: {`opsz`: font size}.

This patch implements the third bucket, which wasn&apos;t implemented in the previous patch. The third
bucket doesn&apos;t actually work correctly on older OSes, so it&apos;s guarded behind a OS version check.

<a href="https://bugs.webkit.org/show_bug.cgi?id=252592">https://bugs.webkit.org/show_bug.cgi?id=252592</a> is about deleting the compatibility third bucket in
favor of the second bucket: actually enabling `trak` for all text that should have optical sizing
enabled.

Just like 260447@main, the tests for this adds a megabyte of fonts (because we have to keep the
whole release of an OFL font together) so I&apos;m not including the test with this PR, and will upload
another PR later with a custom font that I create from scratch for testing this.

* Source/WTF/wtf/PlatformUse.h:
* Source/WebCore/platform/graphics/cocoa/UnrealizedCoreTextFont.cpp:
(WebCore::UnrealizedCoreTextFont::addAttributesForOpticalSizing):
(WebCore::UnrealizedCoreTextFont::applyVariations):
(WebCore::UnrealizedCoreTextFont::modifyFromContext):
(WebCore::UnrealizedCoreTextFont::realize const):
(WebCore::applyVariations): Deleted.
(WebCore::modifyFromContext): Deleted.
* Source/WebCore/platform/graphics/cocoa/UnrealizedCoreTextFont.h:

Canonical link: <a href="https://commits.webkit.org/260781@main">https://commits.webkit.org/260781@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/47febba0b128d5de957fdd5c137993fa6d09efcf

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109325 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18404 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42137 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/872 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/118535 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/113208 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19847 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9688 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101562 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115080 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/14878 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/98105 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/43072 "Found 1 new test failure: webgl/2.0.y/conformance2/vertex_arrays/vertex-array-object.html (failure)") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/96850 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29758 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/84824 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/98295 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11175 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31101 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/99190 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/9244 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11911 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/8032 "Passed tests") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/46/builds/31051 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/17278 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50704 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/107155 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7459 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13529 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26454 "Passed tests") | 
| | | | |

### CODE REVIEW

**Comment by litherum** [comment]
All these tests pass on Ventura, so I think we need a version check.

File: Source/WTF/wtf/PlatformUse.h [hunk 1]
```diff
@@ -391,3 +391,10 @@
 // The header says "Before macOS 13.0 and iOS 16.0 this attribute is not accurate"
 #define USE_KCTFONTVARIATIONAXESATTRIBUTE 1
 #endif
+
+#if (PLATFORM(MAC) && __MAC_OS_X_VERSION_MAX_ALLOWED >= 130000) \
```
Comment (litherum) [inline Source/WTF/wtf/PlatformUse.h:395]: Update copyright

**Comment by litherum** [comment]
Verified tests pass on Ventura

**Comment by litherum** [comment]
This patch should be disabled on Monterey.

**Review by alanbaradlay — APPROVED** [review]

**Pull request merged**